File: backend/app/api/routes/events.py
"""Server-sent events: the live half of the API.

One long-lived GET per browser tab. The stream carries notifications only — a
name and a row id — and the client answers each by re-fetching through the
ordinary routes. So the data path is unchanged, still tenant-scoped by
TenantScope and still auditable; only the "when" moved.

Why SSE and not a websocket: the traffic is one-directional and the client has
nothing to say back. SSE is a plain HTTP response, so it inherits the CORS
config, the auth dependency and the error handlers already in place, and the
browser reconnects on its own.

Why not EventSource on the client: it cannot set an Authorization header, and
the alternative — a token in the query string — puts a credential into access
logs, proxy logs and browser history. The frontend reads this with `fetch` and
a ReadableStream instead; see frontend/services/events.ts.
"""
from __future__ import annotations

import asyncio
import logging
from typing import AsyncIterator

from fastapi import APIRouter
from fastapi.responses import StreamingResponse

from app.api.deps import TenantDep
from app.events.broker import Event, broker

logger = logging.getLogger(__name__)

router = APIRouter(tags=["events"])

# Idle streams look dead to intermediaries. Nginx and most managed proxies drop
# a connection after 30-60s of silence, so send a comment frame well inside
# that. A comment is the cheapest legal SSE frame and clients ignore it.
_HEARTBEAT_SECONDS = 15.0

# Streams end themselves, and the reason is authentication. The Clerk token was
# verified when the request arrived and is never re-checked afterwards — a
# stream held open for a day would be running on a credential that expired
# minutes in. Capping the lifetime means the client reconnects with a fresh
# token, which re-runs the whole auth path. Well under the token's own lifetime
# would be pointless churn; well over it is the thing being avoided.
_MAX_STREAM_SECONDS = 600.0


def _frame(event: Event) -> bytes:
    # No JSON encoder here on purpose: entity_id is a UUID from our own
    # database and name is a module-level constant, so there is nothing to
    # escape, and keeping the frame this literal makes it obvious that no row
    # data can reach it.
    return f"event: {event.name}\ndata: {{\"id\": \"{event.entity_id}\"}}\n\n".encode()


async def _stream(doctor_id: str) -> AsyncIterator[bytes]:
    loop = asyncio.get_running_loop()
    deadline = loop.time() + _MAX_STREAM_SECONDS

    with broker.subscribe(doctor_id) as queue:
        # Tells the client the stream is live. It fetches its initial state on
        # seeing this, which also closes the gap between subscribing and the
        # first event — anything that changed while connecting is picked up by
        # that fetch rather than missed.
        yield b": connected\n\n"

        while True:
            remaining = deadline - loop.time()
            if remaining <= 0:
                # Named so the client reconnects deliberately rather than
                # treating a clean close as an error.
                yield b"event: reconnect\ndata: {}\n\n"
                return

            try:
                event = await asyncio.wait_for(
                    queue.get(), timeout=min(_HEARTBEAT_SECONDS, remaining)
                )
            except asyncio.TimeoutError:
                yield b": ping\n\n"
                continue

            yield _frame(event)


@router.get("/events")
async def stream_events(scope: TenantDep) -> StreamingResponse:
    """Subscribe to the caller's own change notifications.

    Scoped by the token like every other route: TenantScope is constructed from
    the verified subject, and the broker is keyed on the same value, so a
    stream cannot be pointed at another tenant.
    """
    return StreamingResponse(
        _stream(scope.doctor_id),
        media_type="text/event-stream",
        headers={
            # no-transform matters as much as no-cache: compressing proxies
            # buffer, and a buffered event stream is not a live one.
            "Cache-Control": "no-cache, no-transform",
            "Connection": "keep-alive",
            # Nginx-specific, ignored elsewhere. Without it nginx buffers the
            # response and nothing arrives until the stream ends.
            "X-Accel-Buffering": "no",
        },
    )

File: frontend/services/events.ts
// Live change notifications from the backend.
//
// The server sends a name and a row id, never the row — so receiving an event
// means "re-fetch", not "here is the new state". That is what keeps the data
// path unchanged: every read still goes through the ordinary API, where it is
// tenant-scoped and auditable. See backend/app/events/broker.py.
//
// Not EventSource. It cannot set an Authorization header, and the usual
// workaround — the token in a query string — writes a credential into access
// logs, proxy logs and browser history. A streaming `fetch` can send the
// header, at the cost of parsing the SSE framing here, which is a dozen lines.

import { API_URL, authorizedFetch } from './api';

export type ClarusEvent = {
  /** The SSE event type, e.g. "call_log.updated". */
  name: string;
  /** The row that changed. */
  id: string;
};

// The server closes each stream after ten minutes so the next one re-verifies
// a fresh Clerk token, and that close is a normal end rather than a failure —
// hence a short first retry. The backoff is for the other case: a backend that
// is down should not be hammered by every open tab.
const INITIAL_RETRY_MS = 1_000;
const MAX_RETRY_MS = 30_000;

/**
 * Subscribe to the signed-in doctor's own events. Returns an unsubscribe
 * function; call it on unmount or the stream outlives the component.
 */
export function subscribeToEvents(onEvent: (event: ClarusEvent) => void): () => void {
  let closed = false;
  let controller: AbortController | null = null;
  let timer: ReturnType<typeof setTimeout> | null = null;
  let retryDelay = INITIAL_RETRY_MS;

  async function connect(): Promise<void> {
    if (closed) return;

    controller = new AbortController();
    try {
      const response = await authorizedFetch(`${API_URL}/api/events`, {
        signal: controller.signal,
        headers: { Accept: 'text/event-stream' },
      });

      if (!response.ok || !response.body) {
        throw new Error(`Event stream failed (${response.status})`);
      }

      // Reset only after a connection actually succeeded. Resetting on attempt
      // would turn the backoff into a fixed 1s poll against a dead backend.
      retryDelay = INITIAL_RETRY_MS;
      await readFrames(response.body, onEvent);
    } catch (error) {
      // An abort is us calling unsubscribe, not a failure.
      if (closed || (error as Error)?.name === 'AbortError') return;
    }

    if (closed) return;
    timer = setTimeout(connect, retryDelay);
    retryDelay = Math.min(retryDelay * 2, MAX_RETRY_MS);
  }

  void connect();

  return () => {
    closed = true;
    controller?.abort();
    if (timer) clearTimeout(timer);
  };
}

async function readFrames(
  body: ReadableStream<Uint8Array>,
  onEvent: (event: ClarusEvent) => void,
): Promise<void> {
  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';

  for (;;) {
    const { done, value } = await reader.read();
    // The server ended the stream — its lifetime cap, or a restart. Either
    // way the caller reconnects.
    if (done) return;

    // stream: true so a multi-byte character split across two chunks is held
    // rather than decoded into a replacement character.
    buffer += decoder.decode(value, { stream: true });

    let boundary = buffer.indexOf('\n\n');
    while (boundary !== -1) {
      const frame = buffer.slice(0, boundary);
      buffer = buffer.slice(boundary + 2);
      const parsed = parseFrame(frame);
      if (parsed) onEvent(parsed);
      boundary = buffer.indexOf('\n\n');
    }
  }
}

function parseFrame(frame: string): ClarusEvent | null {
  let name = 'message';
  let data = '';

  for (const line of frame.split('\n')) {
    // Comment frames. The server sends these as heartbeats, to stop proxies
    // treating an idle stream as a dead one.
    if (line.startsWith(':')) continue;
    if (line.startsWith('event:')) name = line.slice(6).trim();
    else if (line.startsWith('data:')) data += line.slice(5).trim();
  }

  // Bookkeeping, not a change: the server is about to close so the next
  // connection can carry a fresh token. Reconnecting is handled by the stream
  // ending; surfacing it would trigger a pointless re-fetch.
  if (name === 'reconnect' || !data) return null;

  try {
    const payload = JSON.parse(data) as { id?: string };
    return payload.id ? { name, id: payload.id } : null;
  } catch {
    return null;
  }
}
